pkg/core/sync.go

```diff
@@ -0,0 +1,258 @@
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return diag.FromError(err, diag.DATABASE)
+	}
+
+	return nil
+}
+
+func syncTables(ctx context.Context, sta *state.Client, cur, want *state.Provider, resourceTables map[string]*schema.Table) diag.Diagnostics {
+	want.Tables = resourceTableNames(resourceTables)
+	want.Signatures = resourceSignatures(resourceTables)
+
+	var (
+		curTables     map[string][]string
+		curSignatures map[string]string
+	)
+
+	if cur != nil && len(cur.Tables) > 0 { // Old provider with known, valid data
+		curTables = cur.Tables
+		curSignatures = cur.Signatures
+	} else {
+		curTables = want.Tables // Fallback to installed provider tables
+	}
+
+	tx, err := sta.ProviderSync(ctx)
+	if err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+	defer tx.Rollback(ctx) // nolint:errcheck
+
+	// If a single SQL fails, all following commands also fail with "current transaction is aborted"
+	if diags := dropProvider(ctx, tx, want.Registry(), curTables, curSignatures, want.Signatures); diags.HasErrors() {
+		return diags
+	}
+	if diags := installProvider(ctx, tx, resourceTables); diags.HasErrors() {
+		return diags
+	}
+
+	if err := tx.UninstallProvider(ctx, want.Registry()); err != nil {
+		return diag.FromError(fmt.Errorf("uninstall provider failed: %w", err), diag.INTERNAL)
+	}
+
+	if err := tx.InstallProvider(ctx, want); err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return diag.FromError(err, diag.DATABASE)
+	}
+
+	return nil
+}
+
+func dropProvider(ctx context.Context, db execution.QueryExecer, provider registry.Provider, tableNames map[string][]string, curSignatures, wantSignatures map[string]string) diag.Diagnostics {
+	{
+		migTable := fmt.Sprintf("%s_%s_schema_migrations", provider.Source, provider.Name)
+		q := fmt.Sprintf(dropTableSQL, strconv.Quote(migTable))
+		if err := db.Exec(ctx, q); err != nil {
+			return diag.FromError(err, diag.DATABASE, diag.WithSummary("drop table failed"), diag.WithResourceName(migTable))
+		}
+	}
+
+	for name, tables := range tableNames {
+		if curSignatures != nil && wantSignatures != nil && curSignatures[name] != "" && wantSignatures[name] == curSignatures[name] {
+			log.Debug().Str("resource", name).Str("provider", provider.Name).Msg("keeping tables and all data")
+			continue
+		}
+
+		log.Debug().Str("resource", name).Str("provider", provider.Name).Msg("deleting tables and all relations")
+		for _, t := range tables {
+			if err := db.Exec(ctx, fmt.Sprintf(dropTableSQL, strconv.Quote(t))); err != nil {
+				return diag.FromError(err, diag.DATABASE, diag.WithSummary("drop table failed"), diag.WithResourceName(t))
+			}
+		}
+	}
+
+	return nil
+}
+
+func installProvider(ctx context.Context, db execution.QueryExecer, resourceTables map[string]*schema.Table) diag.Diagnostics {
```

Comment:
got confused with the state.InstallProvider, maybe call this createProviderTables or buildProviderTables?

Reply:
dropProviderTables + createProviderTables